### PR TITLE
Gap junction functionality

### PR DESCRIPTION
**Pull request by janhahne**
This pull request contains the entire gap-junction framework. There are two pynest examples included as well as two unittests. The only thing missing is probably a mpi-test of the gap-junction functionality for the testsuit, as this exceeded my knowledge of sli.

### CODE REVIEW

**Comment by janhahne** [comment]
As discussed in the developer meeting this morning I did some benchmark tests of nest 2.8 in comparison to nest 2.8 including this pull request with gap junction functionality

<h5>Microcircuit model (examples/nest/Potjans_2014)

I used two threads on my local workstation computer and chose different network sizes. Each measured time is the average over 15 runs of the corresponding test

>area = 0.1
 >                        | nest-2.8 | nest-2.8-gap
>Average connecting time  |     5.82 |        8.19
>Average simulation time  |    15.36 |       15.73
> 
>area = 0.2
 >                        | nest-2.8 | nest-2.8-gap
>Average connecting time  |    12.14 |       16.71
>Average simulation time  |    40.55 |       41.62
> 
>area = 0.5
>                                          | nest-2.8 | nest-2.8-gap
>Average connecting time  |    34.36 |       45.46
>Average simulation time  |   126.60 |      126.98
> 
>area = 1.0
 >                        | nest-2.8 | nest-2.8-gap
>Average connecting time  |    82.77 |      103.47
>Average simulation time  |   278.96 |      275.89

<h5>(Brunel, 2000) network

I have rerun the maxfilling of the balanced random network model from (Brunel, 2000) on JUQUEEN.

>VP | nest-2.8 | nest-2.8-gap
>512 | 639.40 | 640.83
>1024 | 536.99 | 541.57
>2048 | 485.05 | 489.52
>4096 | 509.39 | 520.16
>8192 | 626.78 | 631.09
>16384 | 767.69 | 766.54

**Comment by heplesser** [comment]
@janhahne Thank you for providing the benchmark results quickly. The JuQueen results look very fine (I suppose the numbers are build and simulation time together?), as do the simulations times from the microcircuit model on your machine. But I am concerned about the significant increase in connecting time for the microcircuit model on your machine. Could you do a few tests with brunel-2000 on your machine to se if that also is affected? And specify whether the JuQueen numbers include building?

**Comment by janhahne** [comment]
@heplesser The JuQueen results are simulation time only, sorry for the inaccuracy. 

I looked up the corresponding building times, which are as follows:
>VP | nest-2.8 | nest-2.8-gap
>512 | 270.46 | 252.11
>1024 | 253.97 | 237.66 
>2048 | 233.77 | 220.86 
>4096 | 242.00 | 231.84 
>8192 | 293.66 | 273.70
>16384 | 346.72 | 329.37 

The building times are actually lower for the gap code. I have no useful explanation why, but I double-checked the numbers of nodes and connections which were created and those are definitely the same for both branches. As seen below this seems to be a JuQueen specific behaviour.

<h5> (Brunel, 2000) network on local workstation computer

I also did a few tests on my local computer. I used 4 threads and averaged the time over 2 runs
>scale = 0.1
 >                        | nest-2.8 | nest-2.8-gap
>Average connecting time  |     0.57 |        0.65
>Average simulation time  |    99.61 |       99.66
> 
>scale = 0.5
 >                        | nest-2.8 | nest-2.8-gap
>Average connecting time  |     3.65 |        3.88
>Average simulation time  |   242.61 |      241.14
> 
>scale = 1.0
 >                        | nest-2.8 | nest-2.8-gap
>Average connecting time  |     9.64 |        9.84
>Average simulation time  |  1482.01 |     1408.51
> 
>scale = 1.5
>                        | nest-2.8 | nest-2.8-gap
>Average connecting time  |    16.71 |       17.20
>Average simulation time  |  1380.81 |     1359.00

The variability of the simulation times here is most likely a result of the small number of test runs.

**Comment by janhahne** [comment]
I found the reason for the changes in the building time:

In the gap junction code 
`hc->push_back( conn );`
 is replaced by 
`hc->add_connector( b_has_primary, conn );`

```cpp
void
add_connector( bool is_primary, ConnectorBase* conn )
{
  if ( is_primary )
  {
    insert( begin() + primary_end_,
      conn ); // if empty, insert (begin(), conn) inserts into the first position
    ++primary_end_;
  }
  else
  {
    push_back( conn );
  }
}
```

This is needed, as the primary and secondary connections are distinguished by a new parameter `primary_end_` which remembers the end of the primary connections. Therefore new primary connections are no more added by `std::vector::push_back` but by `std::vector::insert`. 

As long as there are only primary connections it is possible to add the connections in the primary case by `push_back( conn );` as well instead of using `insert`. So I compiled a version which does so and it turned out that the build times were effected:

>Area = 0.1
>                         | nest-2.8 | nest-2.8-gap-push-back
>Average connecting time  |     5.43 |        5.82
>Average simulation time  |    14.27 |       15.12
>
>Area = 0.2
>                        | nest-2.8 | nest-2.8-gap-push-back
>Average connecting time  |    11.34 |       12.14
>Average simulation time  |    39.43 |       39.70
>
>Area = 0.5
>                         | nest-2.8 | nest-2.8-gap-push-back
>Average connecting time  |    33.58 |       34.78
>Average simulation time  |   128.65 |      129.00

So long story short: The difference in the build time is the result of the different performances of `std::vector::push_back` and `std::vector::insert`. That also explaines the JuQueen results where the times were the other way around - it seems as if with the JuQueen compiler `std::vector::insert` is the faster procedure.

**Comment by heplesser** [comment]
@janhahne Excellent detective work indeed! But I wonder if the root cause of the build-time difference between your computer isn't a slightly different one than a more efficient implementation of `insert()` on JuQueen. After all, `insert()` adds elements in the middle of a vector and thus must move all subsequent elements, which will cost time; this should not be much of a problem for the networks here, though, since they have no secondary connections so there never is anything beyond `primary_end_` that needs moving.

It appears that the `xlC` compiler on JuQueen actually uses the g++ 4.4.7 STL. I assume your local workstation is a Linux box, so you would also be using the GNU STL. Thus, I suspect it is not difference in implementation that causes the differences.

I am wondering if the cause could rather be the number of connections created for each pre-synaptic neuron on each VP. I have a suspicion (and it is pure suspicion, I haven't done the numbers) that connections on JuQueen are spread out so widely that heterogeneous connectors play a very small role, so `insert()` would hardly be called at all.

**Comment by janhahne** [comment]
@heplesser I see your point and agree with you that the `insert()` vs. `push_back()` issue can not really be the explanation for the JuQueen results.

I have rerun all tests from above, with three different branches:
* nest-master - newest NEST version 
* nest-gap_junctions - newest NEST version including this pull request
* nest-gap_modified - newest NEST version including this pull request + used  `push_back()` instead of `insert()`

To be absolutely sure I did a new compilation for all versions and it turns out that the results of the gap_junction branch on my local computer changed in terms of building time!

<h5> Microcircuit, workstation computer</h5>

Averaged over 15 runs
```
Area = 0.1
Average connecting time
                master-O2 |     5.74
                   gap-O2 |     5.36
          gap-modified-O2 |     5.77
Average simulation time
                master-O2 |    15.51
                   gap-O2 |    15.11
          gap-modified-O2 |    15.41

Area = 0.2
Average connecting time
                master-O2 |    11.86
                   gap-O2 |    11.12
          gap-modified-O2 |    11.88
Average simulation time
                master-O2 |    38.78
                   gap-O2 |    38.94
          gap-modified-O2 |    39.06

Area = 0.5
Average connecting time
                master-O2 |    33.72
                   gap-O2 |    31.61
          gap-modified-O2 |    33.80
Average simulation time
                master-O2 |   117.25
                   gap-O2 |   118.81
          gap-modified-O2 |   119.24

Area = 1.0
Average connecting time
                master-O2 |    82.07
                   gap-O2 |    78.15
          gap-modified-O2 |    82.17
Average simulation time
                master-O2 |   274.92
                   gap-O2 |   277.36
          gap-modified-O2 |   275.39
```


<h5> Brunel, workstation computer</h5>

Averaged over 2 runs
```
scale = 0.1
Average connecting time
                master-O2 |     0.54
                   gap-O2 |     0.58
          gap-modified-O2 |     0.57
Average simulation time
                master-O2 |    99.41
                   gap-O2 |    99.66
          gap-modified-O2 |    99.79

scale = 0.5
Average connecting time
                master-O2 |     3.46
                   gap-O2 |     3.73
          gap-modified-O2 |     3.60
Average simulation time
                master-O2 |   248.94
                   gap-O2 |   246.44
          gap-modified-O2 |   248.56

scale = 1.0
Average connecting time
                master-O2 |     9.20
                   gap-O2 |     9.48
          gap-modified-O2 |     9.66
Average simulation time
                master-O2 |  1526.78
                   gap-O2 |  1476.59
          gap-modified-O2 |  1513.55

scale = 1.5
Average connecting time
                master-O2 |    16.13
                   gap-O2 |    16.39
          gap-modified-O2 |    16.70
Average simulation time
                master-O2 |  1404.40
                   gap-O2 |  1388.30
          gap-modified-O2 |  1396.67
```

<h5> Brunel, JuQueen</h5>

Averaged over 3 runs
```
Average connecting time
      VP |  nest-master | nest-gap_junctions | nest-gap_modified
     512 |       273.04 |             248.71 |       263.44
    1024 |       251.51 |             232.44 |       233.73
    2048 |       233.40 |             222.72 |       218.74
    4096 |       245.18 |             229.71 |       227.40
    8192 |       297.85 |             269.00 |       266.49
   16384 |       353.71 |             319.70 |       321.99
Average simulation time
      VP |  nest-master | nest-gap_junctions | nest-gap_modified
     512 |       639.78 |             638.41 |       643.34
    1024 |       536.21 |             537.40 |       537.88
    2048 |       485.25 |             487.72 |       486.95
    4096 |       512.84 |             511.79 |       506.45
    8192 |       628.32 |             617.68 |       611.18
   16384 |       772.32 |             769.46 |       759.37

```

<h4>Summary</h4>
<h5>Workstation computer:</h5> 
Simulation time is similar. Connection time is even lower for the microcircuit model. This is well explained by the `insert()` vs. `push_back()` issue.
<h5>JuQueen:</h5>
Simulation time is similar. Connection time is lower - this is however not due to the `insert()` vs. `push_back()` issue. Number of connections and results indicate the all connections were created correctly. It seems to me that the reasons for that are most likely found on some lower level and can not be assigned to any particular code line.

**Comment by heplesser** [comment]
:+1: for merge in view of the new benchmark data.

**Comment by janhahne** [comment]
@heplesser @jougs @tammoippen From my point of view this is ready for merging now. Please review.

**Comment by heplesser** [comment]
@janhahne @jougs @tammoippen :+1: for merge once my last comment above is fixed, sorry for the oversight on where to put the unittest import in SLI scripts.
